core/src/main/scala/com/raphtory/api/analysis/graphstate/AccumulatorImplementation.scala

```diff
@@ -0,0 +1,224 @@
+  private val valueA: AtomicInteger        = new AtomicInteger(initialValue)
+  override def currentValue: Int           = currentValueA.get()
+
+  override def merge(other: Int): Unit = this += other
+
+  override def reset(): Unit = {
+    if (retainState)
+      valueA.accumulateAndGet(currentValue, op.asJava)
+    else
+      valueA.set(currentValue)
+    currentValueA.set(initialValue)
+  }
+
+  /** Get last accumulated value */
+  override def value: Int = valueA.get()
+
+  /** Add new value to accumulator
+    *
+    * @param newValue Value to add
+    */
+  override def +=(newValue: Int): Unit =
+    currentValueA.accumulateAndGet(newValue, op.asJava)
+}
+
+private class CounterImplementation[T]() extends Counter[T] {
+  var totalCount: Int = 0
+  var counts          = new java.util.concurrent.ConcurrentHashMap[T, Int]()
+
+  override def getCounts: MapView[T, Int] = counts.asScala.view
+  override def largest: (T, Int)          = getCounts.maxBy(_._2)
+
+  override def largest(k: Int): Iterable[(T, Int)] = {
+    val ord = Ordering.by[(T, Int), Int](_._2).reverse
+    counts.asScala.foldLeft(new mutable.PriorityQueue[(T, Int)]()(ord)) { (q, c) =>
+      q.enqueue(c)
+      if (q.size > k) {
+        q.dequeue()
+        q
+      }
+      else q
+    }
+  }
+
+  def +=(newValue: T, c: Int): Unit =
+    counts.compute(
+            newValue,
+            {
+              case (t, c1) => c + c1
+              case _       => c
+            }
+    )
+
+}
+
+private object CounterImplementation {
+  def apply[T]() = new CounterImplementation[T]()
+}
+
+private class CounterAccumulatorImplementation[T](retainState: Boolean)
+        extends AccumulatorImplementation[(T, Int), Counter[T]] {
+  var value: Counter[T]                      = CounterImplementation()
+  var currentValue: CounterImplementation[T] = CounterImplementation()
+
+  override def merge(other: Counter[T]): Unit = {
+    currentValue.totalCount += other.totalCount
+    other.getCounts.foreach{case (tag, count) => currentValue.counts.merge(tag, count, (c1, c2) => c1 + c2)}
+  }
+
+  override def reset(): Unit = {
+    if (retainState)
+      merge(value)
+    val tmp = value.asInstanceOf[CounterImplementation[T]]
+    value = currentValue
+    currentValue = tmp
+    currentValue.totalCount = 0
+  }
+
+  /** Add new value to accumulator
+    *
+    * @param newValue Value to add
+    */
+  override def +=(newValue: (T, Int)): Unit = currentValue +=(newValue._1, newValue._2)
+}
+
+private object CounterAccumulatorImplementation {
+  def apply[T](retainState: Boolean) = new CounterAccumulatorImplementation[T](retainState)
+}
+
+private class HistogramImplementation[T: Numeric](
+    val noBins: Int,
+    override val minValue: T,
+    override val maxValue: T
+) extends Histogram[T] {
+  val bins: Array[Int] = Array.fill(noBins)(0)
+  var totalCount: Int  = 0
+
+  override def getBins: Array[Int] = bins
+
+  override def cumSum: Array[Int] = bins.scanLeft(0)(_ + _)
```

Comment:
Weird name...

Reply:
that's not me :laughing:

Reply:
😳

Reply:
We found him! :laughing: 

https://github.com/Raphtory/Raphtory/blob/development/core/src/main/scala/com/raphtory/api/analysis/graphstate/Histogram.scala#L35